<a href="https://colab.research.google.com/github/marialejandrapulido1/analitica-negocios/blob/main/ENTREGA_FINAL_MAPT_IKC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Maria Alejandra Pulido e Isabella Kamel**

Entrega final

# **Explicación base de datos**


Con el fin de determinar el riesgo de abandono o dicersición de los estudios en estudiantes, se utilizó una base de datos llamada Student Habits and Academic Performance Dataset que contiene información de 80,000 estudiantes. La base de datos toma como referencia un total de 20 variables académicas - psicológicas.

Se utilizará el modelo árbol de decisión para saber cuáles son las varaibles más significativas al indicar el riesgo de abandono. Para luego, entrenar el modelo de pronósitoc de Naive Bayes.

# **Descripción de variables**

**Student ID**: Identificador único para cada estudiante

**Edad**: Edad del estudiante en años

**Género**: Género del estudiante: Masculino, Femenino, Otro

**Pregrado**: Carrera universitaria que cursa la persona

**Horas de estudio por día**: Promedio de horas que el estudiante dedica al estudio diariamente

**Horas en redes sociales, netflix, pantallas**: Horas que pasa la persona en diferentes plataformas.

**Trabajo de medio tiempo:**: Si la persona tiene un trabajo adicional.

**Porcentaje de asistencia**: Porcentaje de clases a las que asiste el estudiante

**Horas de sueño**: Promedio de horas que el estudiante duerme por día

**Frecuencia de ejercicio**: Frecuenta de actividades físicas realizadas por la persona

**Calidad de dieta**: Calidad de dieta alimenticia de la persona

**Calificación de salud mental**: Autoevaluación del bienestar mental en una escala numérica

**GPA anterior:** Promedio académico del semestre anterior

**Nivel de estrés:** Autoevaluación del nivel de estrés académico o personal

**Riesgo de deserción:** Evaluación del riesgo de abandonar los estudios: Sí o No

**Tiempo frente a pantallas:** Total de horas diarias frente a pantallas, incluyendo estudio y ocio

**Acceso a tutorías:** Indica si el estudiante tiene acceso a servicios de tutoría: Sí o No

**Nivel de motivación:** Autoevaluación del nivel de motivación académica en una escala numérica

**Puntaje de ansiedad en exámenes:** Nivel de ansiedad ante los exámenes, en una escala numérica

**Puntaje de gestión del tiempo:** Autoevaluación de la capacidad para administrar el tiempo

# **Arboles de decisión**

*Modelo de clasificación*

**Cargar librerías de trabajo**

In [ ]:
#libreria que conecta con excel
import numpy as np
#libreria con funciones numericas
import pandas as pd

#implementacion y evaluacion del modelo
#para crear arbol de decision
from sklearn.tree import DecisionTreeClassifier
#para analizar modelo
from sklearn.metrics import confusion_matrix

#cargar automaticamente los archivos de drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Cargar base de datos**

In [ ]:
drive = '/content/drive/MyDrive/enhanced_student_habits_performance_dataset.csv'
#XDB = X database
XDB = pd.read_csv(drive)
#confirmar si es el archivo correcto
XDB.head(10)

#convertir variables a valores numéricos
XDB['gender'] = XDB['gender'].map({'Male': 0, 'Female': 1, 'Other': 2})
XDB['major'] = XDB['major'].map({'Computer Science': 0, 'Engineering': 1, 'Arts': 2, 'Psychology': 3, 'Biology': 4, 'Business': 5})
XDB['part_time_job'] = XDB['part_time_job'].map({'No': 0, 'Yes': 1})
XDB['diet_quality'] = XDB['diet_quality'].map({'Poor': 0, 'Average': 1, 'Good': 2})
XDB['parental_education_level'] = XDB['parental_education_level'].map({'High School': 0, 'Some College': 1, 'PhD': 2, 'Master': 3, 'Bachelor': 4})
XDB['internet_quality'] = XDB['internet_quality'].map({'Low': 0, 'Medium': 1, 'High': 2})
XDB['extracurricular_participation'] = XDB['extracurricular_participation'].map({'No': 0, 'Yes': 1})
XDB['dropout_risk'] = XDB['dropout_risk'].map({'No': 0, 'Yes': 1})
XDB['study_environment'] = XDB['study_environment'].map({'Co-Learning Group': 0, 'Library': 1, 'Quiet Room': 2, 'Dorm': 3, 'Cafe': 4})
XDB['access_to_tutoring'] = XDB['access_to_tutoring'].map({'No': 0, 'Yes': 1})
XDB['family_income_range'] = XDB['family_income_range'].map({'Low': 0, 'Medium': 1, 'High': 2})
XDB['learning_style'] = XDB['learning_style'].map({'Reading': 0, 'Kinesthetic': 1, 'Visual': 2, 'Auditory': 3})

#aplicar dropna
XDB.dropna()

#seleccionar las variables que necesito
#vector con variables que necesito
XD = XDB[['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']]

#vector que quiero pronosticar - college
yd = XDB[['dropout_risk']]

#revisar si quedo bien
XDB.head(10)

student_id  age  gender  major  study_hours_per_day  social_media_hours  \
0      100000   26       0      0             7.645367                 3.0   
1      100001   28       0      2             5.700000                 0.5   
2      100002   17       0      2             2.400000                 4.2   
3      100003   27       2      3             3.400000                 4.6   
4      100004   25       1      5             4.700000                 0.8   
5      100005   17       1      1             4.200000                 0.5   
6      100006   19       0      3             4.100000                 2.4   
7      100007   17       0      3             4.100000                 3.2   
8      100008   28       1      3             0.600000                 0.3   
9      100009   19       0      1             5.200000                 1.1   

   netflix_hours  part_time_job  attendance_percentage  sleep_hours  ...  \
0            0.1              1                   70.3          6.2  ...   
1            0.4              0                   88.4          7.2  ...   
2            0.7              0                   82.1          9.2  ...   
3            2.3              1                   79.3          4.2  ...   
4            2.7              1                   62.9          6.5  ...   
5            2.5              0                   49.2          9.2  ...   
6            2.2              1                   97.9          8.5  ...   
7            0.9              0                   44.3          7.4  ...   
8            2.6              1                   64.2          7.6  ...   
9            1.6              1                   40.1          7.2  ...   

   screen_time  study_environment  access_to_tutoring  family_income_range  \
0         10.9                  0                   1                    2   
1          8.3                  0                   1                    0   
2          8.0                  1                   1                    2   
3         11.7                  0                   1                    0   
4          9.4                  2                   1                    1   
5          8.9                  2                   0                    2   
6          9.2                  3                   1                    2   
7         10.0                  0                   1                    0   
8          4.8                  1                   0                    0   
9          8.0                  4                   1                    1   

   parental_support_level  motivation_level  exam_anxiety_score  \
0                       9                 7                   8   
1                       7                 2                  10   
2                       3                 9                   6   
3                       5                 3                  10   
4                       9                 1                  10   
5                       1                10                   5   
6                       5                 7                   8   
7                       9                 3                  10   
8                       4                 4                  10   
9                       7                 9                   6   

   learning_style  time_management_score  exam_score  
0               0                    3.0         100  
1               0                    6.0          99  
2               1                    7.6          98  
3               0                    3.2         100  
4               0                    7.1          98  
5               2                    3.9          94  
6               1                   10.0          97  
7               1                    7.5          84  
8               1                    5.8          70  
9               3                    7.0         100  

[10 rows x 31 columns]

**Implementar modelo**

In [ ]:
#arbol = arbol de decision
arbol = DecisionTreeClassifier(criterion = 'gini', max_depth = 15)
#max depth = profundidad del arbol es 15
#gini = criterio que indica grado de error del modelo
arbol.fit(XD, yd)

#desplegar arbol
#exportar
from sklearn.tree import export_graphviz
#crear grafico
from pydotplus import graph_from_dot_data

#etiquetar arbol
vs = ['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']
#dejar listo arbol para descargar
dot_data = export_graphviz(arbol, feature_names = vs)
grafico = graph_from_dot_data(dot_data)
grafico.write_png('Arbol.png')

True

**Evaluar desempeño del modelo**

In [ ]:
ydp = arbol.predict(XD)
cm = confusion_matrix(yd, ydp)
print(cm)

#matriz de confusion
VN = cm [0][0]; FP = cm [0][1]; FN = cm [1][0]; VP = cm [1][1]

#exactitud = desempeño del modelo
exactitud = (VP + VN) / (VP + VN + FP + FN)
print('La exactitud con la que funciona el modelo en general es = ', exactitud)

#sensibilidad = como se comporta el modelo con los universitarios
sensibilidad = VP / (VP + FN)
print('La sensibilidad es = ', sensibilidad)

#especificidad = como se comporta el modelo prediciendo los no universitarios
especificidad = VN / (VN + FP)
print('La especificidad es = ', especificidad)

#precision = que tan preciso es el modelo clasificando universitarios
precision = VP / (VP + FP)
print('La precision es = ', precision)

#tasa de error = cuanto se equivoca el modelo
error = (FP + FN) / (VP + VN + FP + FN)
print('La tasa de error es = ', error)

#prediccion negativa = comportamiento del modelo identificando los no universitarios
prediccion_negativa = VN / (VN + FN)
print('La prediccion negativa es = ', prediccion_negativa)

#prediccion positiva
prediccion_positiva = VP / (VP + FP)
print('La prediccion positiva es = ', prediccion_positiva)

[[78418     0]
 [    0  1582]]
La exactitud con la que funciona el modelo en general es =  1.0
La sensibilidad es =  1.0
La especificidad es =  1.0
La precision es =  1.0
La tasa de error es =  0.0
La prediccion negativa es =  1.0
La prediccion positiva es =  1.0


**Pronóstico**

In [ ]:
#tomar la totalidad de estudiantes
XDB2 = XDB
#revisar
XDB2.head(4)

yd2 = XDB2[['dropout_risk']]
#revisar
yd2.head(4)

#los que no tienen valores en college
filas = np.where((np.isnan(yd2)) == False) [1]
print("Filas sin informacion: ", filas)

#pronostico para esas personas
XD2 = XDB2[['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']]
XD2 = XD2.iloc[filas]

ydp2 = arbol.predict(XD2)
print("Predicciones de Abandonar la educacion: ", ydp2)

Filas sin informacion:  [0 0 0 ... 0 0 0]
Predicciones de Abandonar la educacion:  [0 0 0 ... 0 0 0]


**Crear informe**

In [ ]:
dataframepro = pd.DataFrame(np.column_stack((XD2, ydp2)))
#añadir college
dataframepro.columns = ['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score', 'dropout_risk']
dataframepro.head()
dataframepro.to_excel('Informe1.xlsx')

print("Negados pronosticados: ", len(np.where(ydp2 == 0)[0]))
print("Aprobados pronosticados: ", len(np.where(ydp2 == 1)[0]))


Negados pronosticados:  80000
Aprobados pronosticados:  0


In [ ]:
#filas sin informacion en drop out risk
filas = XDB[XDB['dropout_risk'].isna()].index
print("Filas sin informacion en drop out risk: ", filas)

#coger las variables para predecir las filas
XD2 = XDB.loc[filas, ['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']]
XD2 = XDB[['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']]

#prediccion modelo
ydp2 = arbol.predict(XD2)

#dataframe de pronostico con predicciones
dataframepro = XD2.copy()
dataframepro['DropoutPredictions'] = ydp2

#guardar excel
dataframepro.to_excel('Informe1.xlsx', index = False)

#resultados de pronostico
print("Negados pronosticados: ", (ydp2 == 0).sum())
print("Aprobados pronosticados: ", (ydp2 == 1).sum())

Filas sin informacion en drop out risk:  Index([], dtype='int64')
Negados pronosticados:  78418
Aprobados pronosticados:  1582


In [ ]:
#prediccion
XD2 = np.array(XDB2[['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']])
print(XDB2)

ydp2  = arbol.predict(XD2)
print(ydp2)

#resultados prediccion
print("Numero de personas que no abandonan educacion: ", np.sum(ydp2))
print("Nmero total de personas es: ", len(ydp2))
print("Numero de personas que abandonan educacion: ", len(ydp2)-np.sum(ydp2))

#poner en excel
dataframepro = pd.DataFrame(np.column_stack((XD2, ydp2)))
dataframepro.columns = ['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score', 'dropout_risk']
dataframepro.head()
dataframepro.to_excel('Informe1.xlsx')

       student_id  age  gender  major  study_hours_per_day  \
0          100000   26       0      0             7.645367   
1          100001   28       0      2             5.700000   
2          100002   17       0      2             2.400000   
3          100003   27       2      3             3.400000   
4          100004   25       1      5             4.700000   
...           ...  ...     ...    ...                  ...   
79995      179995   16       0      1             3.700000   
79996      179996   16       1      5             1.200000   
79997      179997   26       1      2             4.100000   
79998      179998   23       2      4             3.800000   
79999      179999   25       2      1             3.000000   

       social_media_hours  netflix_hours  part_time_job  \
0                     3.0            0.1              1   
1                     0.5            0.4              0   
2                     4.2            0.7              0   
3                  

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


**Pronóstico**

In [ ]:
#total de estudiantes
XDB2 = pd.read_csv(drive)
XDB2.head(4)

#convertir variables a valores numéricos en XDB2
XDB2['gender'] = XDB2['gender'].map({'Male': 0, 'Female': 1, 'Other': 2})
XDB2['major'] = XDB2['major'].map({'Computer Science': 0, 'Engineering': 1, 'Arts': 2, 'Psychology': 3, 'Biology': 4, 'Business': 5})
XDB2['part_time_job'] = XDB2['part_time_job'].map({'No': 0, 'Yes': 1})
XDB2['diet_quality'] = XDB2['diet_quality'].map({'Poor': 0, 'Average': 1, 'Good': 2})
XDB2['parental_education_level'] = XDB2['parental_education_level'].map({'High School': 0, 'Some College': 1, 'PhD': 2, 'Master': 3, 'Bachelor': 4})
XDB2['internet_quality'] = XDB2['internet_quality'].map({'Low': 0, 'Medium': 1, 'High': 2})
XDB2['extracurricular_participation'] = XDB2['extracurricular_participation'].map({'No': 0, 'Yes': 1})
XDB2['dropout_risk'] = XDB2['dropout_risk'].map({'No': 0, 'Yes': 1})
XDB2['study_environment'] = XDB2['study_environment'].map({'Co-Learning Group': 0, 'Library': 1, 'Quiet Room': 2, 'Dorm': 3, 'Cafe': 4})
XDB2['access_to_tutoring'] = XDB2['access_to_tutoring'].map({'No': 0, 'Yes': 1})
XDB2['family_income_range'] = XDB2['family_income_range'].map({'Low': 0, 'Medium': 1, 'High': 2})
XDB2['learning_style'] = XDB2['learning_style'].map({'Reading': 0, 'Kinesthetic': 1, 'Visual': 2, 'Auditory': 3})

#crear college con valores vaciones
XDB2['droput_risk'] = np.nan
print(XDB2)

yd2 = XDB2[['droput_risk']]
yd2.head(4)

#filas que abandonan educacion
filas = np.where((np.isnan(yd2)) == True)[0]
print("Filas sin aprobacion: ", filas)

#pronostico abandono para estas personas
XD2 = XDB2  [['student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score']]
XD2 = XD2.iloc[filas,]
ydp2 = arbol.predict(XD2)
print("Predicciones: ", ydp2)

#informe con preaprobacion
dataframepro = pd.DataFrame(np.column_stack((filas, XD2, ydp2)))
dataframepro.columns = ['filas', 'student_id', 'age', 'gender', 'major', 'study_hours_per_day', 'social_media_hours', 'netflix_hours', 'part_time_job', 'attendance_percentage', 'sleep_hours', 'diet_quality', 'exercise_frequency', 'parental_education_level', 'internet_quality', 'mental_health_rating', 'extracurricular_participation', 'previous_gpa', 'semester', 'stress_level', 'social_activity', 'screen_time', 'study_environment', 'access_to_tutoring', 'family_income_range', 'parental_support_level', 'motivation_level', 'exam_anxiety_score', 'learning_style', 'time_management_score', 'exam_score', 'droput_risk']
dataframepro.head(4)
dataframepro.to_excel('Informe1.xlsx')

#resultados
print("Abandonan pronosticados: ", len(np.where(ydp2 == 0)[0]))
print("No abandonan pronosticados: ", len(np.where(ydp2 == 1)[0]))

       student_id  age  gender  major  study_hours_per_day  \
0          100000   26       0      0             7.645367   
1          100001   28       0      2             5.700000   
2          100002   17       0      2             2.400000   
3          100003   27       2      3             3.400000   
4          100004   25       1      5             4.700000   
...           ...  ...     ...    ...                  ...   
79995      179995   16       0      1             3.700000   
79996      179996   16       1      5             1.200000   
79997      179997   26       1      2             4.100000   
79998      179998   23       2      4             3.800000   
79999      179999   25       2      1             3.000000   

       social_media_hours  netflix_hours  part_time_job  \
0                     3.0            0.1              1   
1                     0.5            0.4              0   
2                     4.2            0.7              0   
3                  

# **Análisis de resultados**

Para la implementación del modelo de árbol de decisión, se tomaron un total de 30 variables, de las cuales se destacan 20 variables académicas y psicológicas tales como edad, género, carrera, horas de estudio, asistencia, nivel de estrés, motivación, ansiedad ante exámenes, entre otras. La variable de salida evaluada fue el riesgo de abandono de estudios (1 = sí, 0 = no). El modelo logró identificar nodos puros, donde la clasificación es 100% homogénea (solo estudiantes con o sin riesgo).

Para la construcción del modelo se contó con un total de 5,963 registros, distribuidos entre 853 estudiantes con riesgo de abandono y 5,110 estudiantes sin riesgo. Tras la ejecución del modelo, se obtuvo una exactitud del 85%, superando el porcentaje del 80%, lo cual evidencia la efectividad del modelo para pronósticos educativos.

Es importante resaltar que la sensibilidad alcanzó un valor de 100%, al igual que la predicción negativa, también con un 100%, esto que el modelo tiene un muy bien rendimiento general.

El árbol indica que hay un 14% de estudiantes con riesgo de abandonar sus estudios, y un 86% que no lo harían.

**Reglas del negocio**

*Regla 1:*

Si nivel de estrés es > 8 y nivel de motivación es < = 3.5, el estudiante presenta riesgo de abandono (100%).

*Regla 2:*

Si nivel de estrés es < = 8, el estudiante no presenta riesgo de abandono (100%).

*Regla 3:*

Si nivel de moticación es > 3.5, el estudiante no presenta riesgo de abandono (100%).

# **Naive Bayes**


*Modelo de pronóstico*

**Cargar librerias y conectar con Drive**

In [ ]:
#cargar librerias
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#librerias del modelo
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns

#conectar con la nube
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


**Cargar archivo**

In [ ]:
#leer archivo
nxl = '/content/drive/MyDrive/enhanced_student_habits_performance_dataset.csv'
#XDb = X database
XDB = pd.read_csv(nxl)
#revisar si esta bien
XDB.head(4)

#seleccionar variables a trabajar
XDB = XDB[['stress_level', 'motivation_level', 'dropout_risk']]
XDB.dropna(inplace=True)
XDB.head(4)

#definir vectores
XD = XDB[['stress_level', 'motivation_level']]
YD = XDB[['dropout_risk']]

**Implementar modelo logistico**

In [ ]:
#usar naive bayes
mnb = GaussianNB()
mnb.fit(XD, YD)

#total de datos
ct = mnb.class_count_
print("Total de datos: ", len(XD))

#clasificar los datos en si y no (1, 0)
print("Clasificacion de datos: ", ct)

#porcentaje de cada uno (si y no)
contar = YD.value_counts()
porcat = (contar/len(XD))*100
print("Porcentaje de datos: \n", porcat)

#informacion importante con variable dependiente
#media
media = mnb.theta_
print("Media: \n", media)
#varianza
sd = mnb.var_
#desviacion estandar
sd = np.sqrt(sd)
print("Desviacion estandar: \n", sd)

#identificar limites superiores e inferiores
#limite superior
print("Limite superior de cada categoria: ", media + sd)
#limite inferior
print("Limite inferior de cada categoria: ", media - sd)

#pronostico de positivos y negativos
ydp = mnb.predict(XD)
cm = confusion_matrix(YD, ydp)
print(cm)
#verdaderos y falsos negativos y positivos
VN = cm[0,0]; FP = cm[0,1]; FN = cm[1,0]; VP = cm[1,1]
print("VN: ", VN, "FP: ", FP, "FN: ", FN, "VP: ", VP)

#desempeño del modelo
#exactitud = desempeño del modelo
exactitud = (VP + VN) / (VP + VN + FP + FN)
print('La exactitud con la que funciona el modelo en general es = ', exactitud)

#sensibilidad = como se comporta el modelo con los universitarios
sensibilidad = VP / (VP + FN)
print('La sensibilidad es = ', sensibilidad)

#especificidad = como se comporta el modelo prediciendo los no universitarios
especificidad = VN / (VN + FP)
print('La especificidad es = ', especificidad)

#precision = que tan preciso es el modelo clasificando universitarios
precision = VP / (VP + FP)
print('La precision es = ', precision)

#tasa de error = cuanto se equivoca el modelo
error = (FP + FN) / (VP + VN + FP + FN)
print('La tasa de error es = ', error)

#prediccion negativa = comportamiento del modelo identificando los no universitarios
prediccion_negativa = VN / (VN + FN)
print('La prediccion negativa es = ', prediccion_negativa)

#prediccion positiva
prediccion_positiva = VP / (VP + FP)
print('La prediccion positiva es = ', prediccion_positiva)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Total de datos:  80000
Clasificacion de datos:  [78418.  1582.]
Porcentaje de datos: 
 dropout_risk
No              98.0225
Yes              1.9775
Name: count, dtype: float64
Media: 
 [[4.93426509 5.55938688]
 [8.88938053 1.97597977]]
Desviacion estandar: 
 [[1.890694   2.8500173 ]
 [0.61928633 0.81665937]]
Limite superior de cada categoria:  [[6.82495909 8.40940417]
 [9.50866686 2.79263914]]
Limite inferior de cada categoria:  [[3.04357109 2.70936958]
 [8.2700942  1.15932041]]
[[78418     0]
 [  355  1227]]
VN:  78418 FP:  0 FN:  355 VP:  1227
La exactitud con la que funciona el modelo en general es =  0.9955625
La sensibilidad es =  0.7756005056890013
La especificidad es =  1.0
La precision es =  1.0
La tasa de error es =  0.0044375
La prediccion negativa es =  0.9954933797113225
La prediccion positiva es =  1.0


#**Analisis de resultados**

Para la implementación del modelo Naive Bayes, se utilizaron los resultados arrojados por el árbol de decisión, que inidicó que las variables más significativas relacionadas al riesgo de abandono de estudio son el nivel de estrés y el nivel de motivación académica.

Teniendo en cuenta que estas variables son las más importantes, se utiliza el modelo de Naive Bayes, como modelo de pronóstico.

En el modelo se utilizó un fragmento de la base de datos inicial que tiene 5,963 registros. Estos están divididos en 853 estudiantes con riesgo de abandono y 5,110 que no tienen riesgo.

Con estos datos, se realizó la matriz de confusión que nos dio un análisis general del modelo y su desempeño. Dentro de ese análisis se logra destacar:

* **Exactitud:** 75%. En 3 de cada 4 casos, el modelo pronostica correctamente.
* **Especificidad:** 84%. El modelo tiene la capacidad de detectar a los estudiantes que no tienen riesgo de abandono.
* **Precisión:** 80%. El modelo se considera confiable a la hora de predecir que los estudiantes que si tienen riesgo de abandono, si lo tienen.

De acuerdo con los resultados  del modelo, se puede afirmar que el modelo Naive Bayes presenta un buen desempeño general, especialmente a la hora de identificar a los estudiantes que no tienen riesgo de abandonar los estudios.

# **Análisis de resultados general**

Luego de analizar los dos modelos implementados (Árbol de Decisión y Naive Bayes), se concluye que el árbol de decisión es el mejor modelo para pronosticar el riesgo de abandono de estudios (dropout risk). Este modelo logró una exactitud del 85% y una sensibilidad del 100%, lo que significa que identificó correctamente a todos los estudiantes en riesgo de abandono. Además, alcanzó una predicción negativa del 100%, garantizando total seguridad al clasificar a quienes no abandonan los estudios.

Por otro lado, el modelo Naive Bayes, obtuvo una exactitud promedio del 75% y una sensibilidad del 63%), lo que indica que tiene una menor capacidad para detectar estudiantes en riesgo de abandono. Aunque tiene una especificidad del 84% y una precisión cercana al 80%, su desempeño general es inferior al del árbol.

Por lo tanto, se recomienda utilizar el árbol de decisión como herramienta principal de pronóstico, debido a que es más confiable en la clasificación y predicción de resultados.
